### PR TITLE
Add "next gen compiler" blog post

### PR DESCRIPTION
**Pull request by vmarkovtsev**

### CODE REVIEW

File: content/post/next_compiler.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+Software is a form of human communication; software corpora have similar statistical properties
+to natural language corpora; and these properties can be exploited to build better software
+engineering tools. 
+
+This hypothesis fully expresses what we think and even checked several times (
+[topic modeling](https://arxiv.org/abs/1704.00135),
+[identifier embeddings](https://blog.sourced.tech/post/id2vec/),
+[code prediction](https://github.com/src-d/code-completion)). It is taken from
+["A Survey of Machine Learning for Big Code and Naturalness"](https://arxiv.org/pdf/1709.06182.pdf)
+by Allamanis et.al. This paper is an excellent overview of what we have today in MLoSC, and
```
[the remark this answers is not in the document]
Comment (campoy) [reply content/post/next_compiler.md]: and even better ... mlosc means love in Polish!

[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply content/post/next_compiler.md]: Yep, we kind of invented this term and force it everywhere now.

File: static/post/next_compiler/system.dot [hunk 2]
```diff
@@ -0,0 +1,7 @@
+digraph Language {
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply static/post/next_compiler/system.dot]: It is the Graphviz source of the PNG file.

**Comment by vmarkovtsev** [comment]
@campoy, @eiso is rewriting this post ATM

File: content/post/next_compiler.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+happened many times before in the history.
+
+## Feedback
+
+Programming languages evolve, as the natural ones do. The rules of a natural language reflect
+the norms which are adopted by the majority of the speakers. There is evidence that the complexity
+of a natural language decreases over time. If you compare Russian language from 19th century
+to the one in 21st, you will surprise how ridiculously complex is the former compared to the
+latter: more letters, more rules, more exceptions. Funnily enough, we do not know any programming
+language which does becomes simpler or easier to learn over time.
```
Comment (campoy) [inline content/post/next_compiler.md]: s/becomes/become/

I could argue that Go is an effort to create a language simpler that its predecessors, but your point is still valid

File: content/post/next_compiler.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+There are several reasons we can imagine for that. While in a natural language words are either
```
Comment (campoy) [inline content/post/next_compiler.md]: There are several causes we can image that justify this situation.

File: content/post/next_compiler.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+dropped or replaces, programming languages keep gaining new functions yet at the same time do
```
Comment (campoy) [inline content/post/next_compiler.md]: s/replaces/replaced/

Comment (campoy) [reply content/post/next_compiler.md]: s/functions/features/ to avoid ambiguity

Comment (campoy) [reply content/post/next_compiler.md]: "yet at the same time do not remove old ones and maintain backwards compatibility"

"while keeping all the previous features for backwards compatibility"

File: content/post/next_compiler.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+not remove old ones and maintain backwards compatibility. The sad example of Python 3 clearly
```
Comment (campoy) [inline content/post/next_compiler.md]: Backwards compatibility is expensive but paramount, as the effect of Python's migration from version 2 to 3 has shown.

Avoid using "sad", it's an unncessary judgement call

File: content/post/next_compiler.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+shows that revolutionary changes harm so much that they can split the community.
+
+Francesc, the hed of our Developer Relations, presented a curious example of Go language evolution
```
Comment (campoy) [inline content/post/next_compiler.md]: s/hed of our Developer Relations/VP of Developer Relations at source{d}/

Comment (campoy) [reply content/post/next_compiler.md]: Also, please add a link to my twitter https://twitter.com/francesc from my name.

File: content/post/next_compiler.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+in his [post on Medium](https://medium.com/@francesc/source-d-why-i-left-google-71240c0f0204).
+The community ran the analysis over the existing open source projects and
```
Comment (campoy) [inline content/post/next_compiler.md]: Actually it wasn't the community as much as it was just me spending money on BigQuery 😄 

"Francesc analyzed the existing open source projects and ..."

File: content/post/next_compiler.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+[proved](https://github.com/golang/go/issues/14595) the usefulness of a new function in the
+standard library. It demonstrates how the programming languages develop today: people give
```
Comment (campoy) [inline content/post/next_compiler.md]: It demonstrates how programming languages evolve today: users give

File: content/post/next_compiler.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+explicit feedback (RFCs, JSRs, PEPs, GitHub issues, etc.) and the owners decide on them and
+sometimes fulfill the requests. The time to live is months, if not years. Let's consider how
+we are able change it by applying ML on "big code".
+
+We are not being extreme at first. We can run [pattern mining](https://arxiv.org/pdf/1602.05012.pdf)
+to further automate the proposals process. There are multiple ways to mine, will it be a
```
Comment (campoy) [inline content/post/next_compiler.md]: There are multiple ways to mine, recurrent neural networks with attention and pattern memory being a notable one.

File: content/post/next_compiler.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+recurrent neural network with attention and pattern memory or anything else. It also depends on
+the target pattern shape: fixed and rigid or flexible and template-like small or big.
```
Comment (campoy) [inline content/post/next_compiler.md]: template-like; small or big

File: content/post/next_compiler.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+We bet that the following code will be #1 top recurring snippet in Go:
+
+```go
+..., err := /* whatever func() */
+if err != nil {
+  panic(err)  // or [return] whatever func(err)
```
Comment (campoy) [inline content/post/next_compiler.md]: troll :trollface:

File: content/post/next_compiler.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+}
+```
+
+Examples: [here](https://gobyexample.com/reading-files)
```
Comment (campoy) [inline content/post/next_compiler.md]: I'd replace the "here" with something meaningful:

We've found this pattern while [reading files](...), ...

File: content/post/next_compiler.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+or [here](https://github.com/divan/gotrace/blob/master/main.go#L38)
+or [here](https://github.com/moby/moby/blob/master/client/client.go#L141)
+or [here](https://github.com/src-d/go-git/blob/master/_examples/common.go#L19)
+(We can easily continue this for a long time). Yet this is a feature of the initial language design
```
Comment (campoy) [inline content/post/next_compiler.md]: Drop the "(We can easily continue this for a long time)"

File: content/post/next_compiler.md [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+and no way its owners are doing anything with it. This is an important observation: even if the
```
Comment (campoy) [inline content/post/next_compiler.md]: This is way too negative 👎 

Unfortunately, changing the way errors are handled in Go doesn't seem to be a priority for the Go community, and any change at this point would probably break backwards compatibility.

File: content/post/next_compiler.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+pattern is statistically obvious, it may be rejected by a tiny group of people who invented
+the language and are maintaining it. By no means we blame anyone; this is their right.
```
Comment (campoy) [inline content/post/next_compiler.md]: Please, do not write this. Attacking the maintainers of an open source project should not be our style.

File: content/post/next_compiler.md [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+People are always biased and professionally deformed. By definition, open source projects
```
Comment (campoy) [inline content/post/next_compiler.md]: "Professionally deformed" makes me cringe. I know there's a French expression for this which has been also adopted by Spanish culture ... but this is not a thing people want to read in a technical document.

File: content/post/next_compiler.md [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+first solve the authors' problems and only then everybody else's. The question is,
+can we design a language and the ecosystem around it to completely exclude the human factor?
+With Machine Learning, we can.
+
+The principle of democracy is to make decisions which are supported by the majority. If we had
+a way to not only mine the patterns but also vote for the optimization suggestions and have them
+automatically applied in the next release, we would dramatically reduce the lags and the controversy.
+Each language user in isolation cannot always know if his code is typical for many other developers
```
Comment (campoy) [inline content/post/next_compiler.md]: Language users in isolation cannot always know whether their code is typical for many other developers or not.

**Comment by campoy** [comment]
And after sending my review I see that @eiso is working on it ... oh well! 🤦‍♂️

**Comment by eiso** [comment]
Still appreciate it. But yes, I need to find more time (I will) to rework
this post. Ideas in it are valuable but this doesn't tell the right story
yet.

In a way this post feels to me like the ingredients to a good black mirror
episode, programming language edition.

File: content/post/next_compiler.md (hunk 17 above)
Comment (vmarkovtsev) [reply content/post/next_compiler.md]: Haha, ok. Copied from Russian.

**Comment by vmarkovtsev** [comment]
@campoy Thanks for such a long a detailed review! I will learn much from it.

Regarding the Go attack, I think I struggled with some Go code back then, reviewed PRs, communicated with the ML internship candidates and wrote the post at the same, which made me... stressed.

Awesome point about designing a new language: we can say that we fork Go and break the backwards compatibility :)

Backwards compatibility: since the patterns are inferred automatically, I would expect an automatically evolving conversion tool to exist.

**Comment by vmarkovtsev** [comment]
> I wonder whether rather than showing how programming languages should evolve with MLoSC we should talk about how they could be designed on the first place extracting lessons from existing programming languages and MLoSC.

The idea here is to make a living language, which is constantly designed. For sure the initial design matters but it is going to be patched in the future iterations.

**Comment by campoy** [comment]
I'm reviewing content for the blog post for the following months.
Could we publish this on March 13th?

**Comment by eiso** [comment]
It requires the rewrite I promised Vadim a long time ago. March 13th is feasible

**Comment by vmarkovtsev** [comment]
@eiso Ping

**Comment by vmarkovtsev** [comment]
Noooooooooooooooo I force pushed and erased everything, my reflog is empty and the local branch was not synced. @eiso How about the rewrite :trollface:

**Comment by eiso** [comment]
Ouch, and yeah, we'll have to take this idea from the top. I have a draft written somewhere, unfortunately, no idea where. This is on me. I still have this one locally btw if you want me to push it.

**Comment by marnovo** [comment]
#GitEpicFails

**Pull request closed without merging**